Source/RetryPolicy.swift

```diff
@@ -0,0 +1,379 @@
+    ///   - retryLimit:               The total number of times the request is allowed to be retried. `2` by default.
+    ///   - exponentialBackoffBase:   The base of the exponential backoff policy. `2` by default.
+    ///   - exponentialBackoffScale:  The scale of the exponential backoff. `0.5` by default.
+    ///   - retryableHTTPMethods:     The HTTP methods that are allowed to be retried.
+    ///                               `RetryPolicy.defaultRetryableHTTPMethods` by default.
+    ///   - retryableHTTPStatusCodes: The HTTP status codes that are automatically retried by the policy.
+    ///                               `RetryPolicy.defaultRetryableHTTPStatusCodes` by default.
+    ///   - retryableURLErrorCodes:   The URL error codes that are automatically retried by the policy.
+    ///                               `RetryPolicy.defaultRetryableURLErrorCodes` by default.
+    public init(
+        retryLimit: UInt = RetryPolicy.defaultRetryLimit,
+        exponentialBackoffBase: UInt = RetryPolicy.defaultExponentialBackoffBase,
+        exponentialBackoffScale: Double = RetryPolicy.defaultExponentialBackoffScale,
+        retryableHTTPMethods: Set<HTTPMethod> = RetryPolicy.defaultRetryableHTTPMethods,
+        retryableHTTPStatusCodes: Set<Int> = RetryPolicy.defaultRetryableHTTPStatusCodes,
+        retryableURLErrorCodes: Set<URLError.Code> = RetryPolicy.defaultRetryableURLErrorCodes)
+    {
+        precondition(exponentialBackoffBase >= 2, "The `exponentialBackoffBase` must be a minimum of 2.")
+
+        self.retryLimit = retryLimit
+        self.exponentialBackoffBase = exponentialBackoffBase
+        self.exponentialBackoffScale = exponentialBackoffScale
+        self.retryableHTTPMethods = retryableHTTPMethods
+        self.retryableHTTPStatusCodes = retryableHTTPStatusCodes
+        self.retryableURLErrorCodes = retryableURLErrorCodes
+    }
+}
+
+// MARK: -
+
+extension RetryPolicy: RequestRetrier {
+    public func should(
+        _ session: Session,
+        retry request: Request,
+        with error: Error,
+        completion: @escaping (_ result: Result<TimeInterval>) -> Void)
+    {
+        if
+            request.retryCount < retryLimit,
+            let httpMethod = request.request?.method,
+            retryableHTTPMethods.contains(httpMethod),
+            isRetryable(request.response, with: error)
+        {
+            let timeDelay = pow(Double(exponentialBackoffBase), Double(request.retryCount)) * exponentialBackoffScale
+            completion(.success(timeDelay))
+        } else {
+            completion(.failure(error))
+        }
+    }
+
+    private func isRetryable(_ response: HTTPURLResponse?, with error: Error) -> Bool {
+        if let statusCode = response?.statusCode, retryableHTTPStatusCodes.contains(statusCode) {
+            return true
+        } else if let errorCode = (error as? URLError)?.code, retryableURLErrorCodes.contains(errorCode) {
+            return true
+        }
```

Comment:
@jshier moving these into extensions makes things worse IMO since the status codes and error codes can be customized. Seemed best IMO to keep this logic private and rely on the instance state of the retry policy.

Reply:
Renamed in b9d0b5d7.